File: samples/objective-c/ios/compressed-streams/CompressedStreamsSample/CompressedStreamsSample/ViewController.m
//
// Copyright (c) Microsoft.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

#import "ViewController.h"
#import <AVFoundation/AVFoundation.h>
#import <MicrosoftCognitiveServicesSpeech/SPXSpeechApi.h>


@interface ViewController () {
    NSString *speechKey;
    NSString *serviceRegion;
}

@property (strong, nonatomic) IBOutlet UIButton *recognizeFromMP3FileButton;

@property (strong, nonatomic) IBOutlet UILabel *recognitionResultLabel;

- (IBAction)recognizeFromMP3FileButtonTapped:(UIButton *)sender;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    speechKey = @"YourSubscriptionKey";
    serviceRegion = @"YourServiceRegion";

    [self.view setBackgroundColor:[UIColor whiteColor]];

    self.recognizeFromMP3FileButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.recognizeFromMP3FileButton addTarget:self action:@selector(recognizeFromMP3FileButtonTapped:) forControlEvents:UIControlEventTouchUpInside];
    [self.recognizeFromMP3FileButton setTitle:@"Recognize from MP3 file" forState:UIControlStateNormal];
    [self.recognizeFromMP3FileButton setFrame:CGRectMake(50.0, 100.0, 300.0, 50.0)];
    self.recognizeFromMP3FileButton.accessibilityIdentifier = @"recognize_mic_button";
    [self.view addSubview:self.recognizeFromMP3FileButton];

    self.recognitionResultLabel = [[UILabel alloc] initWithFrame:CGRectMake(50.0, 350.0, 300.0, 400.0)];
    self.recognitionResultLabel.lineBreakMode = NSLineBreakByWordWrapping;
    self.recognitionResultLabel.numberOfLines = 0;
    self.recognitionResultLabel.accessibilityIdentifier = @"result_label";
    [self.recognitionResultLabel setText:@"Press a button!"];

    [self.view addSubview:self.recognitionResultLabel];
}

- (IBAction)recognizeFromMP3FileButtonTapped:(UIButton *)sender {
    dispatch_async(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
        [self recognizeFromMP3File];
    });
}

/*
 * Performs speech recognition from a MP3 file.
 */
- (void)recognizeFromMP3File {
    NSBundle *mainBundle = [NSBundle mainBundle];
    NSString *weatherFile = [mainBundle pathForResource: @"whatstheweatherlike" ofType:@"mp3"];
    if (!weatherFile) {
        NSLog(@"Cannot find audio file!");
        [self updateRecognitionErrorText:(@"Cannot find audio file")];
        return;
    }

    // <setup-stream>
    SPXAudioStreamContainerFormat compressedStreamFormat = SPXAudioStreamContainerFormat_MP3;
    SPXAudioStreamFormat *audioFormat = [[SPXAudioStreamFormat alloc] initUsingCompressedFormat:compressedStreamFormat];
    SPXPushAudioInputStream* stream = [[SPXPushAudioInputStream alloc] initWithAudioFormat:audioFormat];

    SPXAudioConfiguration* audioConfig = [[SPXAudioConfiguration alloc] initWithStreamInput:stream];
    if (!audioConfig) {
        NSLog(@"Error creating stream!");
        [self updateRecognitionErrorText:(@"Error creating stream!")];
        return;
    }
    // </setup-stream>

    SPXSpeechConfiguration *speechConfig = [[SPXSpeechConfiguration alloc] initWithSubscription:speechKey region:serviceRegion];
    if (!speechConfig) {
        NSLog(@"Could not load speech config");
        [self updateRecognitionErrorText:(@"Speech Config Error")];
        return;
    }

    SPXSpeechRecognizer* speechRecognizer = [[SPXSpeechRecognizer alloc] initWithSpeechConfiguration:speechConfig audioConfiguration:audioConfig];
    if (!speechRecognizer) {
        NSLog(@"Could not create speech recognizer");
        [self updateRecognitionResultText:(@"Speech Recognition Error")];
        return;
    }

    // connect callbacks
    [speechRecognizer addRecognizingEventHandler: ^ (SPXSpeechRecognizer *recognizer, SPXSpeechRecognitionEventArgs *eventArgs) {
        NSLog(@"Received intermediate result event. SessionId: %@, recognition result:%@. Status %ld. offset %llu duration %llu resultid:%@", eventArgs.sessionId, eventArgs.result.text, (long)eventArgs.result.reason, eventArgs.result.offset, eventArgs.result.duration, eventArgs.result.resultId);
        [self updateRecognitionStatusText:eventArgs.result.text];
    }];


    [speechRecognizer addRecognizedEventHandler: ^ (SPXSpeechRecognizer *recognizer, SPXSpeechRecognitionEventArgs *eventArgs) {
        NSLog(@"Received final result event. SessionId: %@, recognition result:%@. Status %ld. offset %llu duration %llu resultid:%@", eventArgs.sessionId, eventArgs.result.text, (long)eventArgs.result.reason, eventArgs.result.offset, eventArgs.result.duration, eventArgs.result.resultId);
        [self updateRecognitionResultText:eventArgs.result.text];
    }];

    // <push-compressed-stream>
    NSInputStream *compressedStream = [[NSInputStream alloc] initWithFileAtPath:weatherFile];
    [compressedStream open];
    NSLog(@"result of opening stream: %@, %lu", compressedStream.streamError, (unsigned long)compressedStream.streamStatus);
    if (nil == compressedStream)
    {
        NSLog(@"Error while opening file");
        audioConfig = nil;
        return;
    }

    // start recognizing
    [self updateRecognitionStatusText:(@"Recognizing from push stream...")];
    [speechRecognizer startContinuousRecognition];

    const NSInteger nBytesToRead = 1000;
    // push data to stream;
    uint8_t *buffer = malloc(nBytesToRead);
    NSInteger nBytesRead = 0;
    while (1)
    {
        // read data
        nBytesRead = [compressedStream read:buffer maxLength:nBytesToRead];
        if (0 == nBytesRead) {
            NSLog(@"end of stream reached");
            [stream close];
            break;
        }
        else if (0 > nBytesRead) {
            NSLog(@"error reading stream (%ld): %@ (%lu)", nBytesRead, compressedStream.streamError, compressedStream.streamStatus);
            [stream close];
            break;
        }
        else
        {
            NSLog(@"Read %lu bytes from file", nBytesRead);
            NSData *data = [NSData dataWithBytesNoCopy:buffer length:nBytesRead freeWhenDone:NO];

            [stream write:data];
            NSLog(@"Wrote %lu bytes to stream", [data length]);
        }

        [NSThread sleepForTimeInterval:0.1f];
    }

    [speechRecognizer stopContinuousRecognition];
    // </push-compressed-stream>
}

- (void)updateRecognitionResultText:(NSString *) resultText {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.recognitionResultLabel.textColor = UIColor.blackColor;
        self.recognitionResultLabel.text = resultText;
    });
}

- (void)updateRecognitionErrorText:(NSString *) errorText {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.recognitionResultLabel.textColor = UIColor.redColor;
        self.recognitionResultLabel.text = errorText;
    });
}

- (void)updateRecognitionStatusText:(NSString *) statusText {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.recognitionResultLabel.textColor = UIColor.grayColor;
        self.recognitionResultLabel.text = statusText;
    });
}

@end

File: samples/swift/ios/compressed-streams/CompressedStreamsSample/ViewController.swift
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

import UIKit
import MicrosoftCognitiveServicesSpeech

class ViewController: UIViewController {
    var label: UILabel!
    var recognizeButton: UIButton!

    let speechKey = "YourSubscriptionKey"
    let serviceRegion = "YourServiceRegion"

    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = .white

        recognizeButton = UIButton(frame: CGRect(x: 50, y: 100, width: 300, height: 50))
        recognizeButton.setTitle("Recognize from MP3 file", for: .normal)
        recognizeButton.setTitleColor(.systemBlue, for: .normal)
        recognizeButton.accessibilityIdentifier = "recognize_mp3_button"
        recognizeButton.addTarget(self, action: #selector(recognizeButtonTapped), for: .touchUpInside)
        view.addSubview(recognizeButton)

        label = UILabel(frame: CGRect(x: 50, y: 200, width: 300, height: 400))
        label.textColor = .gray
        label.lineBreakMode = .byWordWrapping
        label.numberOfLines = 0
        label.accessibilityIdentifier = "result_label"
        label.text = "Press a button!"
        view.addSubview(label)
    }

    @objc func recognizeButtonTapped() {
        DispatchQueue.global(qos: .default).async {
            self.recognizeFromMP3File()
        }
    }

    /*
     * Performs speech recognition from an MP3 file using a push stream.
     */
    func recognizeFromMP3File() {
        guard let weatherFilePath = Bundle.main.path(forResource: "whatstheweatherlike", ofType: "mp3") else {
            print("Cannot find audio file!")
            updateLabel(text: "Cannot find audio file", color: .red)
            return
        }

        // <setup-stream>
        let audioFormat = SPXAudioStreamFormat(usingCompressedFormat: .mp3)!
        let pushStream = SPXPushAudioInputStream(audioFormat: audioFormat)!
        let audioConfig = SPXAudioConfiguration(streamInput: pushStream)!
        // </setup-stream>

        guard let speechConfig = try? SPXSpeechConfiguration(subscription: speechKey, region: serviceRegion) else {
            print("Could not load speech config")
            updateLabel(text: "Speech Config Error", color: .red)
            return
        }

        guard let recognizer = try? SPXSpeechRecognizer(speechConfiguration: speechConfig, audioConfiguration: audioConfig) else {
            print("Could not create speech recognizer")
            updateLabel(text: "Speech Recognition Error", color: .red)
            return
        }

        recognizer.addRecognizingEventHandler { [weak self] _, evt in
            print("Recognizing: \(evt.result.text ?? "")")
            self?.updateLabel(text: evt.result.text, color: .gray)
        }

        recognizer.addRecognizedEventHandler { [weak self] _, evt in
            print("Recognized: \(evt.result.text ?? "")")
            self?.updateLabel(text: evt.result.text, color: .black)
        }

        // <push-compressed-stream>
        guard let inputStream = InputStream(fileAtPath: weatherFilePath) else {
            print("Error opening audio file")
            updateLabel(text: "Error opening audio file", color: .red)
            return
        }
        inputStream.open()

        updateLabel(text: "Recognizing from push stream...", color: .gray)
        recognizer.startContinuousRecognition()

        let bufferSize = 1000
        let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: bufferSize)
        defer { buffer.deallocate() }

        while true {
            let bytesRead = inputStream.read(buffer, maxLength: bufferSize)
            if bytesRead == 0 {
                print("End of stream reached")
                pushStream.close()
                break
            } else if bytesRead < 0 {
                print("Error reading stream: \(inputStream.streamError?.localizedDescription ?? "unknown")")
                pushStream.close()
                break
            } else {
                let data = Data(bytes: buffer, count: bytesRead)
                pushStream.write(data)
            }
            Thread.sleep(forTimeInterval: 0.1)
        }

        recognizer.stopContinuousRecognition()
        // </push-compressed-stream>
    }

    func updateLabel(text: String?, color: UIColor) {
        DispatchQueue.main.async {
            self.label.text = text
            self.label.textColor = color
        }
    }
}
